osu.Game/Rulesets/Mods/ModFlashlight.cs

```diff
@@ -1,17 +1,158 @@
+            rulesetContainer.KeyBindingInputManager.Add(flashlight);
+
+            flashlight.Breaks = rulesetContainer.Beatmap.Breaks;
+        }
+
+        public abstract Flashlight CreateFlashlight();
+
+        public abstract class Flashlight : Drawable
+        {
+            internal BindableInt Combo;
+            private Shader shader;
+
+            protected override DrawNode CreateDrawNode() => new FlashlightDrawNode();
+
+            public override bool RemoveCompletedTransforms => false;
+
+            public List<BreakPeriod> Breaks;
+
+            protected override void ApplyDrawNode(DrawNode node)
+            {
+                base.ApplyDrawNode(node);
+
+                var flashNode = (FlashlightDrawNode)node;
+
+                flashNode.Shader = shader;
+                flashNode.ScreenSpaceDrawQuad = ScreenSpaceDrawQuad;
+                flashNode.MousePosWrapper.FlashlightPosition = Vector2Extensions.Transform(FlashlightPosition, DrawInfo.Matrix);
+                flashNode.MousePosWrapper.FlashlightSize = Vector2Extensions.Transform(FlashlightSize, DrawInfo.Matrix);
+            }
+
+            [BackgroundDependencyLoader]
+            private void load(ShaderManager shaderManager)
+            {
+                shader = shaderManager.Load("PositionAndColour", FragmentShader);
+            }
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+
+                Combo.ValueChanged += OnComboChange;
+
+                this.FadeInFromZero(FLASHLIGHT_FADE_DURATION);
+
+                foreach (var breakPeriod in Breaks)
+                {
+                    if (breakPeriod.Duration < FLASHLIGHT_FADE_DURATION * 2) continue;
+
+                    this.Delay(breakPeriod.StartTime + FLASHLIGHT_FADE_DURATION).FadeOutFromOne(FLASHLIGHT_FADE_DURATION);
+                    this.Delay(breakPeriod.EndTime - FLASHLIGHT_FADE_DURATION).FadeInFromZero(FLASHLIGHT_FADE_DURATION);
+                }
+            }
+
+            protected abstract void OnComboChange(int newCombo);
+
+            protected abstract string FragmentShader { get; }
+
+            private Vector2 flashlightPosition;
+            protected Vector2 FlashlightPosition
+            {
+                get => flashlightPosition;
+                set
+                {
+                    if (flashlightPosition == value) return;
+
+                    flashlightPosition = value;
+                    Invalidate(Invalidation.DrawNode);
+                }
+            }
+
+            private Vector2 flashlightSize;
+            protected Vector2 FlashlightSize
+            {
+                get => flashlightSize;
+                set
+                {
+                    if (flashlightSize == value) return;
+
+                    flashlightSize = value;
+                    Invalidate(Invalidation.DrawNode);
+                }
+            }
+        }
+
+        public struct FlashlightUniformWrapper
```

Comment:
Still necessary? Might as well merge it into the drawnode tbh, it's a bit awkward that it's called "MousePosWrapper", and "UniformWrapper" when it's really neither of those.